Add support of Job Names with folders (issue #210) + Add JobDSL plugin to the Vanilla Docker image

addihorn opened the pull request:
allows parameter ```--job-name``` to create a job inside a folder-structure.
The parameter will be splitted at ```/``` to allow creation and assignment of jobs to specific folders.

Folders can be created inside ```Runner``` (if not created otherwise) or by Job DSL to allow introduction of additional library-settings etc.

addihorn committed:
Add support of job names for folders

parameter ```--job-name``` will be splitted at ```/``` to allow creation and assignme of jobs to specific folders
folders will be created either by JFR-Engine (If not existing) or by Job DSL to allow introduction of additional library-settings etc.

oleg-nenashev committed:
Merge branch 'master' into feature/jobFolderSupport

oleg-nenashev reviewed (commented):
I am fine with adding Johb DSL to the vanilla package, as long as it is not a mandatory dependency. Actually it is even better since we get support for job definitions in JCasC. Maybe it is another mode to be added later: triggering job from JCasC or Groovy hooks instead of passing Jenkinsfile.  I will create a follow-up ticket, no need to do it here unless you want to do it shortly.

@addihorn before merging, could you please align the codestyle with the existing codebase?

E.g. 

```java
if (condition) {
 // ...
}
```

addihorn commented:
hi @oleg-nenashev 

Thanks for the review.
Just to understand you correctly
>I am fine with adding Johb DSL to the vanilla package, as long as it is not a mandatory dependency.

You want me, to mark the JobDSL as optional dependency in the vanilla package pom, or is it okay, since the plugin is not referenced in payloads?

oleg-nenashev commented:
Sorry for confusion. The current version looks good to me function-wise. Just needs some codestyle cleanup. Happy to merge/release after that

addihorn committed:
Alligned Codestyle with existing code base

+ defined correct Dockerfile to build development-systemn
+ fixed volume mounts in job-folder examples

addihorn commented:
Thank you for the clarification. Just wanted to make sure.

Alligned the code with your existing code base, as you asked.

oleg-nenashev commented on demo/job-folders/README.md:
```diff
@@ -40,15 +42,15 @@ docker run --rm \
 docker run --rm \
 	-v $(pwd)/demo/job-folders/Jenkinsfile:/workspace/Jenkinsfile \
-	jenkins4eval/jenkinsfile-runner:dev \
+	enkins4eval/jenkinsfile-runner:dev \
```
```suggestion
	jenkins4eval/jenkinsfile-runner:dev \
```

typo

oleg-nenashev commented on payload/src/main/java/io/jenkins/jenkinsfile/runner/Runner.java:
```diff
@@ -106,6 +125,22 @@ public int run(PipelineRunOptions runOptions) throws Exception {
+
+            Folder folder = j.getItem(folderName, addToFolder.getItemGroup(), Folder.class);
+            return folder !=null ? folder : addToFolder.createProject(Folder.class, folderName);
+        } catch (IOException ex) {
```
Would be nice to add logging there. This is a wrong state which needs to be reported, and it should lead to the execution failure IMHO

oleg-nenashev commented on demo/job-folders/JenkinsfileWithLibraryCall:
```diff
@@ -0,0 +1,14 @@
+
+
+
+
```
```suggestion
```

oleg-nenashev reviewed (commented):
Would be nice to also address the error propagation before the merge

addihorn committed:
error propagation from method createOrReturnFolder(Folder, String)

In case of an IOException, an Error-Message will be logged to stderr.
The exception is propagated into the calling method and the execution will fail

+ fixed typo as per review

oleg-nenashev reviewed (approved):
Looks good to me. Thanks for addressing the comments @addihorn !
